File: backend/src/services/rag/textChunker.js
function getTailOverlap(text, overlap) {
  if (!text || overlap <= 0) return "";
  return text.slice(Math.max(0, text.length - overlap)).trim();
}

function splitTextIntoChunks(text, options = {}) {
  const chunkSize = Number(
    options.chunkSize || process.env.RAG_CHUNK_SIZE || 1000,
  );
  const overlap = Number(
    options.overlap || process.env.RAG_CHUNK_OVERLAP || 200,
  );

  if (!text || !text.trim()) {
    return [];
  }

  const paragraphs = text
    .split(/\n{2,}/)
    .map((paragraph) => paragraph.trim())
    .filter(Boolean);

  const chunks = [];
  let currentChunk = "";

  for (const paragraph of paragraphs) {
    const candidate = currentChunk
      ? `${currentChunk}\n\n${paragraph}`
      : paragraph;

    if (candidate.length <= chunkSize) {
      currentChunk = candidate;
      continue;
    }

    if (currentChunk) {
      chunks.push(currentChunk.trim());
      const overlapText = getTailOverlap(currentChunk, overlap);
      currentChunk = overlapText ? `${overlapText}\n\n${paragraph}` : paragraph;
    } else {
      currentChunk = paragraph;
    }

    while (currentChunk.length > chunkSize) {
      const slice = currentChunk.slice(0, chunkSize).trim();
      if (slice) {
        chunks.push(slice);
      }
      currentChunk = currentChunk
        .slice(Math.max(0, chunkSize - overlap))
        .trim();
    }
  }

  if (currentChunk.trim()) {
    chunks.push(currentChunk.trim());
  }

  return chunks.map((chunk, chunkIndex) => ({
    chunkIndex,
    text: chunk,
  }));
}

module.exports = {
  splitTextIntoChunks,
};

File: backend/python_rag_service/text_chunker.py
from .config import RAG_CHUNK_OVERLAP, RAG_CHUNK_SIZE


def get_tail_overlap(text: str, overlap: int) -> str:
    if not text or overlap <= 0:
        return ""
    return text[max(0, len(text) - overlap) :].strip()


def split_text_into_chunks(text: str, chunk_size: int | None = None, overlap: int | None = None):
    selected_chunk_size = chunk_size or RAG_CHUNK_SIZE
    selected_overlap = overlap if overlap is not None else RAG_CHUNK_OVERLAP

    if not text or not text.strip():
        return []

    paragraphs = [part.strip() for part in text.split("\n\n") if part.strip()]
    chunks: list[str] = []
    current_chunk = ""

    for paragraph in paragraphs:
        candidate = f"{current_chunk}\n\n{paragraph}" if current_chunk else paragraph
        if len(candidate) <= selected_chunk_size:
            current_chunk = candidate
            continue

        if current_chunk:
            chunks.append(current_chunk.strip())
            overlap_text = get_tail_overlap(current_chunk, selected_overlap)
            current_chunk = f"{overlap_text}\n\n{paragraph}" if overlap_text else paragraph
        else:
            current_chunk = paragraph

        while len(current_chunk) > selected_chunk_size:
            slice_text = current_chunk[:selected_chunk_size].strip()
            if slice_text:
                chunks.append(slice_text)
            current_chunk = current_chunk[max(0, selected_chunk_size - selected_overlap) :].strip()

    if current_chunk.strip():
        chunks.append(current_chunk.strip())

    return [{"chunkIndex": index, "text": value} for index, value in enumerate(chunks)]
